Add docs for variables to xarray attributes #393

tomwhite opened the pull request:
This is a draft to show how it might work. I haven't converted all the docstrings yet, but will do if this looks OK.

Here is an example rendered dataset - note the `comment` attributes (they are hidden by default and can be viewed by clicking on the file icon):

![ds](https://user-images.githubusercontent.com/85085/102084467-1acbe880-3e0d-11eb-90ec-2686d1aebff6.png)

I also checked that the Sphinx generated documentation is the same (note use of `dataclasses.field(repr=False)` so that the docstring does not show up as an attribute in the string representation of `ArrayLikeSpec`).

Fixes #393

tomwhite commented:
Test failures are due to #421

jeromekelleher commented:
Fancy! LGTM

ravwojdyla reviewed (approved):
@tomwhite great, I'm happy this has worked without too much resistance. The only real comment I have is the naming right now. There are two methods:
 * `_check_field` - which used to just check field, now also optionally adds doc
 * `validate` - which used to just "validate" dataset, now might also add doc

I wonder if we should reflect that in the names of the methods?

tomwhite commented:
@ravwojdyla I agree that the names aren't right. I think we might need two functions, since methods typically validate their inputs (which shouldn't modify the input dataset), and validate new variables in their output (which would be a good place to update attributes as well). So we could have `validate` and `validate_and_update_attributes`?

Another option would be to have a utility function that creates (known) variables in a dataset and sets the attributes at the time of creation.

eric-czech commented:
Very cool!

ravwojdyla commented:
> I think we might need two functions, since methods typically validate their inputs (which shouldn't modify the input dataset), and validate new variables in their output (which would be a good place to update attributes as well). So we could have validate and validate_and_update_attributes?

Naming ... 🤔  How about `validate` for to "validate inputs", and `annotate` to "annotate newly created dataset". We can say that `annotate` goes over the variables and validates and annotates them. wdyt @tomwhite?

> Another option would be to have a utility function that creates (known) variables in a dataset and sets the attributes at the time of creation.

afair I've played around with that kind of API, but it ended up being more user code, but I'm curious how would you envision that @tomwhite?

tomwhite commented:
> afair I've played around with that kind of API, but it ended up being more user code, but I'm curious how would you envision that

I was thinking of a wrapper around the `xarray.Dataset` [constructor](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.html) that delegates to Xarray, then sets comment attributes for all registered variables.

ravwojdyla commented:
@tomwhite I'm +1, as long as that is going to play well with `conditional_merge_datasets`/`merge_datasets`, but I guess we can have both custom xarray ctor and public "annotate" method used by the ctor.

tomwhite commented:
The latest commit (https://github.com/pystatgen/sgkit/pull/420/commits/af41aef7687f71d5a4dbb647cfef9f19f9ad3153) shows a `create_dataset` function that calls `annotate`.

tomwhite commented:
@ravwojdyla what do you think of the approach in https://github.com/pystatgen/sgkit/pull/420/commits/af41aef7687f71d5a4dbb647cfef9f19f9ad3153? Shall I apply it across the codebase?

ravwojdyla commented on sgkit/utils.py:
```diff
@@ -194,6 +195,16 @@ def define_variable_if_absent(
+    data_vars: Mapping[Hashable, Any] = None,  # type: ignore[assignment]
+    coords: Mapping[Hashable, Any] = None,  # type: ignore[assignment]
+    attrs: Mapping[Hashable, Any] = None,  # type: ignore[assignment]
+) -> Dataset:
```
\+ doc from xr.Dataset CTOR?

ravwojdyla commented on sgkit/variables.py:
```diff
@@ -115,10 +117,59 @@ def _check_field(
             ) from e
 
+    @classmethod
+    def _annotate(
```
These should probably reuse code from `_validate`/`_check_field`?

ravwojdyla commented:
@tomwhite I'm +1. I do want to highlight one thing tho, this adds a "sgkit blessed" xarray dataset contractor, which is somewhat close to the idea of a lightweight wrapper around xarray, but arguable a bit harder to discover. Also does the implementation in af41aef work well with `conditional_merge_datasets`/`merge_datasets`?

tomwhite commented:
> I do want to highlight one thing tho, this adds a "sgkit blessed" xarray dataset contractor, which is somewhat close to the idea of a lightweight wrapper around xarray, but arguable a bit harder to discover. 

That's a very good point. All of the variable and validation logic is "opt-in", so it should always be possible for users writing statgen methods to just use regular xarray APIs when using sgkit.

Regarding being harder to discover, the function should be documented better and added to the public API when we are happy with it.

> Also does the implementation in [af41aef](https://github.com/pystatgen/sgkit/commit/af41aef7687f71d5a4dbb647cfef9f19f9ad3153) work well with `conditional_merge_datasets`/`merge_datasets`?

I think there are further possible improvements here. I was thinking of making `create_dataset` also call `variables.validate()`. It could also (optionally) do the dataset merging, or we could have another function `create_dataset_from` that takes an input dataset.

ravwojdyla commented:
> I think there are further possible improvements here. I was thinking of making create_dataset also call variables.validate(). It could also (optionally) do the dataset merging, or we could have another function create_dataset_from that takes an input dataset.

Sounds good, I wonder what's the good middle ground between scope creeping this PR and just adding the descriptions in a good enough way. But I leave that up to you. Regarding `create_dataset` vs `create_dataset_from`, we could also have a single `create_dataset` that optionally takes a variable number of Datasets to merge from. So it would be one stop CTOR for sgkit blessed (validated/documented) Datasets.

tomwhite commented:
Updated to remove code duplication between `validate` and `annotate`, and to convert calls in the main codebase from `xr.Dataset` to `create_dataset`. Also added doc for `create_dataset`.

I've left changes to merging for later.

ravwojdyla reviewed (approved):
🎊
